### PR TITLE
Changing Metadata services to abstract classes without non-default constructors

### LINKED ISSUE #1181: ModelBuilder is not dynamically selected by provider in the correct way
ModelBuilder is added as a common service in AddEntityFramework but an attempt is made to specialize in AddAzureTableStorage.

If the ModelBuilder needs to be specialized, then it needs to be dynamically selected as part of the provider services such that multiple providers can register different model differs without one hiding another.

Note that the cross-store functional tests are not currently catching this issue.

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Don't register CompositeEntityKeyFactory and ValueGeneratorSelector in DI
Make ModelBuilderFactory a provider-only service
Refactor EntityServiceCollectionExtensionsTest

Resolves #1181
Part of #1510

### CODE REVIEW

**Comment by ajcvickers** [comment]
Why removing CompositeEntityKeyFactory  and ValueGeneratorSelector from DI?

**Comment by AndriySvyryd** [comment]
They aren't needed in DI

File: src/EntityFramework.Core/Storage/DataStoreServices.cs [hunk 1]
```diff
@@ -51,6 +51,11 @@ public abstract class DataStoreServices
             get { return p => new DbContextService<DataStoreConnection>(() => GetStoreServices(p).Connection); }
         }
 
+        public static Func<IServiceProvider, DbContextService<ModelBuilderFactory>> ModelBuilderFactoryFactory
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Storage/DataStoreServices.cs]: This might need a better name

**Comment by ajcvickers** [comment]
So, your're saying that nobody could ever possibly want to change the implementation of CompositeEntityKeyFactory. Even though it interacts with value readers which can have quirks depending on the provider implementation?

**Comment by AndriySvyryd** [comment]
No, I'm saying that when you want to change CompositeEntityKeyFactory you need to implement your own EntityKeyFactorySource, this is consistent with how you would change the SimpleEntityKeyFactory.

**Comment by ajcvickers** [comment]
I can see that that is consistent. But the SimpleEntityKeyFactory case is due to limitations in what we can resolve in DI due to needing an instance of a generic type for which we don't know the generic type parameter in advance. It seems to me that CompositeEnittyKeyFactory is a service and can be resolved from DI and hence it should be resolved from DI.

**Comment by ajcvickers** [comment]
:shipit:

**Pull request merged**